nifi-nar-bundles/nifi-scripting-bundle/nifi-scripting-processors/src/main/java/org/apache/nifi/script/ScriptRunnerFactory.java

```diff
@@ -0,0 +1,124 @@
+import org.apache.nifi.script.impl.ClojureScriptRunner;
+import org.apache.nifi.script.impl.GenericScriptRunner;
+import org.apache.nifi.script.impl.GroovyScriptRunner;
+import org.apache.nifi.script.impl.JavascriptScriptRunner;
+import org.apache.nifi.script.impl.JythonScriptRunner;
+
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineFactory;
+import javax.script.ScriptException;
+import java.io.File;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.LinkedList;
+import java.util.List;
+
+public class ScriptRunnerFactory {
+
+    private final static ScriptRunnerFactory INSTANCE = new ScriptRunnerFactory();
+
+    private ScriptRunnerFactory() {
+
+    }
+
+    public static ScriptRunnerFactory getInstance() {
+        return INSTANCE;
+    }
+
+    public ScriptRunner createScriptRunner(ScriptEngineFactory scriptEngineFactory, String scriptToRun, String[] modulePaths)
+            throws ScriptException {
+        ScriptEngine scriptEngine = scriptEngineFactory.getScriptEngine();
+        String scriptEngineName = scriptEngineFactory.getLanguageName();
+        if ("Groovy".equals(scriptEngineName)) {
+            return new GroovyScriptRunner(scriptEngine, scriptToRun, null);
+        }
+        if ("python".equals(scriptEngineName)) {
+            return new JythonScriptRunner(scriptEngine, scriptToRun, modulePaths);
+        }
+        if ("Clojure".equals(scriptEngineName)) {
+            return new ClojureScriptRunner(scriptEngine, scriptToRun, null);
+        }
+        if ("ECMAScript".equals(scriptEngineName)) {
+            return new JavascriptScriptRunner(scriptEngine, scriptToRun, null);
+        }
+        return new GenericScriptRunner(scriptEngine, scriptToRun, null);
+    }
+
+    /**
+     * Scans the given module paths for JARs. The path itself (whether a directory or file) will be added to the list
+     * of returned module URLs, and if a directory is specified, it is scanned for JAR files (files ending with .jar).
+     * Any JAR files found are added to the list of module URLs. This is a convenience method for adding directories
+     * full of JAR files to an ExecuteScript or InvokeScriptedProcessor instance, rather than having to enumerate each
+     * JAR's URL.
+     *
+     * @param modulePaths An array of module paths to scan/add
+     * @param log         A logger for the calling component, to provide feedback for missing files, e.g.
+     * @return An array of URLs corresponding to all modules determined from the input set of module paths.
+     */
+    public URL[] getModuleURLsForClasspath(String scriptEngineName, String[] modulePaths, ComponentLog log) {
+
+        if ("Clojure".equals(scriptEngineName)
+                || "Groovy".equals(scriptEngineName)
+                || "ECMAScript".equals(scriptEngineName)) {
+
+            List<URL> additionalClasspath = new LinkedList<>();
+            if (modulePaths != null) {
```

Comment:
May be worth refactoring to use the "fail-fast" / "escape-fast" approach here:
```
if (modulesPaths == null) {
  return new URL[0];
}
```
And so forth. Should result in minimal nesting and easier-to-read code IMO